server/resp/src/main/java/org/infinispan/server/resp/commands/cluster/NODES.java

```diff
@@ -0,0 +1,207 @@
+import org.infinispan.factories.impl.ComponentRef;
+import org.infinispan.manager.CacheManagerInfo;
+import org.infinispan.manager.ClusterExecutor;
+import org.infinispan.manager.EmbeddedCacheManager;
+import org.infinispan.remoting.transport.Address;
+import org.infinispan.security.AuthorizationPermission;
+import org.infinispan.security.actions.SecurityActions;
+import org.infinispan.server.resp.ByteBufferUtils;
+import org.infinispan.server.resp.Resp3Handler;
+import org.infinispan.server.resp.RespCommand;
+import org.infinispan.server.resp.RespErrorUtil;
+import org.infinispan.server.resp.RespRequestHandler;
+import org.infinispan.server.resp.RespServer;
+import org.infinispan.server.resp.commands.Resp3Command;
+import org.infinispan.topology.CacheTopology;
+
+import io.netty.channel.ChannelHandlerContext;
+import net.jcip.annotations.GuardedBy;
+
+/**
+ * `<code>CLUSTER NODES</code>` command.
+ * <p>
+ *  A command that uses the current {@link CacheTopology} and {@link ConsistentHash} to retrieve information about
+ *  the nodes. The response slightly changes with the node executing the command, as it is necessary to identify itself.
+ *  The execution is broadcast for nodes in the topology to identify themselves. So the response are cached according
+ *  to the node that executed the command and the topology.
+ * </p>
+ *
+ * @link <a href="https://redis.io/commands/cluster-nodes/">CLUSTER NODES</a>
+ * @since 15.0
+ * @author José Bolina
+ */
+public class NODES extends RespCommand implements Resp3Command {
+
+   @GuardedBy("this")
+   protected ConsistentHash hash = null;
+
+   @GuardedBy("this")
+   private final Map<Address, CompletionStage<CharSequence>> requestPerOwner = new HashMap<>();
+
+   public NODES() {
+      super(2, 0, 0, 0);
+   }
+
+   @Override
+   public CompletionStage<RespRequestHandler> perform(Resp3Handler handler, ChannelHandlerContext ctx, List<byte[]> arguments) {
+      handler.checkPermission(AuthorizationPermission.ADMIN);
+      AdvancedCache<?, ?> respCache = handler.cache();
+      DistributionManager dm = respCache.getDistributionManager();
+      if (dm == null) {
+         RespErrorUtil.customError("This instance has cluster support disabled", handler.allocator());
+         return handler.myStage();
+      }
+
+      CacheTopology topology = dm.getCacheTopology();
+      ConsistentHash currentCH = topology.getCurrentCH();
+
+      if (currentCH == null) {
+         RespErrorUtil.customError("No consistent hash available", handler.allocator());
+         return handler.myStage();
+      }
+
+      CompletionStage<CharSequence> cs;
+      synchronized (this) {
+         if (!currentCH.equals(hash)) {
+            requestPerOwner.clear();
+            hash = currentCH;
+         }
+         EmbeddedCacheManager ecm = SecurityActions.getEmbeddedCacheManager(respCache);
+         cs = requestPerOwner.computeIfAbsent(ecm.getAddress(), ignore -> requestClusterInformation(handler, ctx, ecm, topology));
```

Comment:
Isn't `ecm.getAddress()` always the same thing irrespective of the consistent hash? Not sure I understand the requestPerOwner map usage.

Reply:
Not always, it depends on the server handling the request. The result is slightly different and includes a `myself` in the returned nodes list. We have a test covering this, too. With two servers, each client connects to one server.

Reply:
Sorry, totally missed the point. Updated now.